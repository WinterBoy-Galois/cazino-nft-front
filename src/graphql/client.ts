import { ApolloClient, InMemoryCache, from, split } from '@apollo/client';
import { WebSocketLink } from '@apollo/client/link/ws';
import { TokenRefreshLink } from 'apollo-link-token-refresh';
import possibleTypes from './possibleTypes.json';
import { getMainDefinition } from '@apollo/client/utilities';
import { HttpLink, defaultDataIdFromObject } from '@apollo/client';
import jwtDecode from 'jwt-decode';
import { getEpoch } from '../common/util/date.util';
import { setContext } from '@apollo/client/link/context';
import { AuthType } from '../state/models/auth.model';
import { appConfig } from '../common/config';

const cache = new InMemoryCache({
  dataIdFromObject(responseObject) {
    switch (responseObject.__typename) {
      default:
        return defaultDataIdFromObject(responseObject);
    }
  },
  addTypename: true,
  possibleTypes,
});

let wsLink: WebSocketLink;

const getApolloClient = (
  onAccessTokenRefresh: (accessToken: string) => void,
  onSignOut: () => void,
  authType: AuthType,
  accessToken?: string
) => {
  cache.reset();

  if (wsLink) {
    (wsLink as any).subscriptionClient.close();
  }

  wsLink = new WebSocketLink({
    uri: `${appConfig.apiBasePathWS}/graphql`,
    options: {
      reconnect: true,
      connectionParams: {
        authToken: accessToken ?? undefined,
      },
    },
  });

  const httpLink = new HttpLink({
    uri: `${appConfig.apiBasePath}/graphql`,
    credentials: 'include',
  });

  const authLink = setContext((_, { headers }) => {
    if (accessToken) {
      return {
        headers: {
          ...headers,
          authorization: `Bearer ${accessToken}`,
        },
      };
    }

    return headers;
  });

  const tokenLink = new TokenRefreshLink({
    accessTokenField: 'accessToken',
    isTokenValidOrUndefined: () => {
      if (accessToken) {
        const { exp } = jwtDecode(accessToken);

        if (exp < getEpoch()) {
          return false;
        }
      } else if (authType === 'SIGNED_IN') {
        return false;
      }

      return true;
    },
    fetchAccessToken: () => {
      return fetch(`${appConfig.apiBasePath}/refresh_tokens`, {
        method: 'POST',
        credentials: 'include',
      });
    },
    handleFetch: accessToken => onAccessTokenRefresh(accessToken),
    handleError: () => onSignOut(),
  });

  const link = split(
    ({ query }) => {
      const definition = getMainDefinition(query);
      return definition.kind === 'OperationDefinition' && definition.operation === 'subscription';
    },
    wsLink,
    authLink.concat(httpLink)
  );

  return new ApolloClient({
    cache,
    link: from([tokenLink, link]),
  });
};

export default getApolloClient;
