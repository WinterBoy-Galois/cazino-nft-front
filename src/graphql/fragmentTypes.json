{
  "__schema": {
    "types": [
      {
        "kind": "UNION",
        "name": "PublicUserOrError",
        "possibleTypes": [{ "name": "PublicUser" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "LoginResultOrError",
        "possibleTypes": [{ "name": "LoginResult" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "UserOrError",
        "possibleTypes": [{ "name": "User" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "SignedInTokenOrError",
        "possibleTypes": [{ "name": "LoginResult" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "BetResultOrError",
        "possibleTypes": [{ "name": "BetResult" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "BooleanOrError",
        "possibleTypes": [{ "name": "GenericBoolean" }, { "name": "GenericErrorArray" }]
      }
    ]
  }
}
