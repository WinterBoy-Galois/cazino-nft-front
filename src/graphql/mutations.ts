import gql from 'graphql-tag';

export const CHANGE_SERVER_SEED = gql`
  mutation ChangeServerSeed {
    changeServerSeed {
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const ACTIVATE_ACCOUNT = gql`
  mutation ActivateAccount($code: String!) {
    activateAccount(code: $code) {
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const SIGN_IN = gql`
  mutation SignIn($email: String, $password: String, $remember: Boolean) {
    signIn(email: $email, password: $password, remember: $remember) {
      ... on LoginResult {
        accessToken
        user {
          id
          username
          avatarUrl
          balance
          email
          isActivated
          depositAddress
          refCode
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const SIGN_UP = gql`
  mutation SignUp(
    $email: String
    $password: String
    $username: String
    $token: String
    $ref: String
  ) {
    registerUser(
      username: $username
      email: $email
      password: $password
      captcha: $token
      ref: $ref
    ) {
      ... on LoginResult {
        accessToken
        user {
          id
          username
          avatarUrl
          balance
          depositAddress
          refCode
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const SIGN_OUT = gql`
  mutation SignOut {
    signOut
  }
`;

export const RESEND_ACTIVATION_CODE = gql`
  mutation ResendActivationCode {
    resendActivationCode {
      ... on GenericBoolean {
        result
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const RESET_PASSWORD = gql`
  mutation ResetPassword($token: String!, $newPassword: String!) {
    resetPassword(uuid: $token, newPassword: $newPassword) {
      ... on LoginResult {
        accessToken
        user {
          id
          username
          avatarUrl
          balance
          depositAddress
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const RECOVER_PASSWORD = gql`
  mutation RecoverPassword($email: String!) {
    forgotPassword(email: $email) {
      ... on GenericBoolean {
        result
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const UPDATE_PASSWORD = gql`
  mutation UpdatePassword($oldPassword: String!, $newPassword: String!) {
    modifyPassword(oldPassword: $oldPassword, newPassword: $newPassword) {
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const UPDATE_PREFERENCES = gql`
  mutation UpdatePreferences($hideUsername: Boolean!, $hideProfit: Boolean!, $hideWager: Boolean!) {
    modifyPreferences(
      hideUsername: $hideUsername
      hideTotalProfit: $hideProfit
      hideTotalWager: $hideWager
    ) {
      ... on User {
        id
        hideUsername
        hideTotalProfit
        hideTotalWager
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const UPDATE_AVATAR = gql`
  mutation updateAvatar($index: Int!) {
    modifyAvatar(index: $index) {
      ... on User {
        id
        avatarUrl
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const MAKE_BET_DICE = gql`
  mutation makeBetDice($betAmount: Float!, $over: Boolean!, $target: Float!) {
    makeBetDice(betAmount: $betAmount, over: $over, target: $target) {
      ... on BetResultDice {
        id
        result
        lucky
        multiplier
        profit
        balance
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const MAKE_BET_CLAMS = gql`
  mutation makeBetClams($betAmount: Float!, $selection: [Int]) {
    makeBetClams(betAmount: $betAmount, selection: $selection) {
      ... on BetResultClams {
        id
        result
        lucky
        multiplier
        profit
        balance
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const MAKE_BET_GOALS = gql`
  mutation makeBetGoals($betAmount: Float!, $difficulty: GoalsDifficulty!) {
    makeBetGoals(betAmount: $betAmount, difficulty: $difficulty) {
      ... on GoalsGameSetup {
        session {
          betId
          betAmount
          currentStep
          selections {
            step
            luckySpots
            selected
          }
          difficulty
          profits {
            step
            multiplier
            profit
          }
          totalProfit {
            multiplier
            profit
          }
          nextProfit {
            multiplier
            profit
          }
          allowNext
          profitCut
        }
        balance
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const ADVANCE_GOALS = gql`
  mutation advanceGoals($betId: ID!, $selection: Int!) {
    advanceGoals(betId: $betId, selection: $selection) {
      ... on GoalsStep {
        id
        totalProfit {
          multiplier
          profit
        }
        nextProfit {
          multiplier
          profit
        }
        nextStep
        profitCut
        allowNext
      }
      ... on GoalsComplete {
        id
        lucky
        profit {
          multiplier
          profit
        }
        balance
        profitCut
        result {
          step
          luckySpots
          selected
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const CASH_OUT_GOALS = gql`
  mutation cashoutGoals($betId: ID!) {
    cashoutGoals(betId: $betId) {
      ... on GoalsComplete {
        id
        lucky
        balance
        profit {
          multiplier
          profit
        }
        profitCut
        result {
          step
          luckySpots
          selected
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const WITHDRAW = gql`
  mutation withdraw($amount: Float!, $address: String!) {
    withdraw(amount: $amount, address: $address) {
      ... on GenericBoolean {
        result
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const CLAIM_FAUCET = gql`
  mutation claimFaucet {
    claimFaucet {
      ... on User {
        balance
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const MAKE_BET_MINES = gql`
  mutation makeBetMines($betAmount: Float!, $mines: Int!) {
    makeBetMines(betAmount: $betAmount, mines: $mines) {
      __typename
      ... on MinesGameSetup {
        session {
          ... on MinesGameSessionSetup {
            betId
            betAmount
            mines
            open
            totalProfit {
              multiplier
              profit
            }
            nextProfit {
              multiplier
              profit
            }
            allowNext
            profitCut
          }
        }
        balance
        maxProfit
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const ADVANCE_MINES = gql`
  mutation advanceMines($betId: ID!, $selection: Int!) {
    advanceMines(betId: $betId, selection: $selection) {
      ... on MinesStep {
        id
        totalProfit {
          multiplier
          profit
        }
        nextProfit {
          multiplier
          profit
        }
        profitCut
        allowNext
        result
      }
      ... on MinesComplete {
        id
        lucky
        balance
        profit {
          multiplier
          profit
        }
        minePositions
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const CASH_OUT_MINES = gql`
  mutation cashoutMines($betId: ID!) {
    cashoutMines(betId: $betId) {
      ... on MinesComplete {
        id
        lucky
        balance
        profit {
          multiplier
          profit
        }
        minePositions
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const CLAIM_BONUS = gql`
  mutation claimBonus($bonusId: ID!) {
    claimBonus(bonusId: $bonusId) {
      ... on User {
        balance
      }
      ... on GenericErrorArray {
        errors {
          code
          message
        }
      }
    }
  }
`;

export const CLAIM_COMMISSION = gql`
  mutation claimCommissions {
    claimCommissions {
      ... on User {
        id
        username
        email
        balance
        hideUsername
        hideTotalProfit
        hideTotalWager
        isActivated
        avatarUrl
        totalWager
        totalProfit
        mostPlayed
        totalBets
        luckyBets
        depositAddress
        refCode
        refCommissions
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;
