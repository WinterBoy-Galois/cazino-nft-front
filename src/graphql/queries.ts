import gql from 'graphql-tag';

export const RECENT_BETS = gql`
  {
    recentBets {
      allBets {
        id
        time
        userid
        username
        gameid
        bet
        profit
        multiplier
      }
      myBets {
        id
        time
        userid
        username
        gameid
        bet
        profit
        multiplier
      }
    }
  }
`;

export const LEADERBOARDS = gql`
  {
    leaderboards {
      daily {
        username
        userid
        wager
        bonus
      }
      weekly {
        username
        userid
        wager
        bonus
      }
      monthly {
        username
        userid
        wager
        bonus
      }
    }
  }
`;

export const USER_INFO = gql`
  query UserInfo($userId: ID) {
    userInfo(id: $userId) {
      __typename
      ... on PublicUser {
        id
        username
        avatarUrl
        totalWager
        totalProfit
        mostPlayed
        totalBets
        luckyBets
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const USER_INFO_AVATAR_URL = gql`
  query UserInfoAvatarUrl($userId: ID) {
    userInfo(id: $userId) {
      __typename
      ... on PublicUser {
        id
        avatarUrl
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const BET_DETAILS = gql`
  query BetDetails($betId: ID) {
    betDetails(id: $betId) {
      ... on BetDetails {
        bet
        profit
        profitCut
        multiplier
        gameResult {
          ... on DiceResult {
            target
            over
            winChance
            resultFloat
          }
          ... on MinesResult {
            mineCount
            minePositions
            open
          }
          ... on GoalsResult {
            difficulty
            selections {
              step
              luckySpots
              selected
            }
          }
          ... on ClamsResult {
            selection
            resultInteger
          }
        }
      }
    }
  }
`;

export const BET_DETAILS_SERVER_SEED = gql`
  query BetDetailsServerSeed($betId: ID) {
    betDetails(id: $betId) {
      ... on BetDetails {
        seedDetails {
          __typename
          ... on SeedDetailsOther {
            serverSeedHash
          }
          ... on SeedDetailsOwn {
            activeGames
            serverSeedHash
            serverSeed
            clientSeed
            nonce
            results
            verificationUrl
          }
        }
      }
    }
  }
`;

export const ME = gql`
  query Me {
    me {
      id
      username
      avatarUrl
      balance
      email
      isActivated
      depositAddress
      refCommissions
      refCode
    }
  }
`;

export const AFF_STATS = gql`
  query affStats {
    affStats {
      bets
      refs
      wager
      commissions
    }
  }
`;

export const ME_STATISTICS_PREFERENCES = gql`
  query Me {
    me {
      id
      totalWager
      totalProfit
      mostPlayed
      totalBets
      luckyBets
      hideUsername
      hideTotalProfit
      hideTotalWager
    }
  }
`;

export const SETUP_CASHIER = gql`
  query SetupCashier {
    setupCashier {
      networkFee
      depositConfirmations
      minWithdraw
    }
    me {
      id
      depositAddress
    }
  }
`;

export const DEPOSITS = gql`
  query Deposits($page: Int, $limit: Int) {
    transactionsDeposit(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        status
        time
        hash
        amount
      }
    }
  }
`;

export const SETUP_DICE = gql`
  query SetupDice {
    setupDice {
      ... on DiceGameSetup {
        minProbability
        maxProbability
        he
      }
    }
  }
`;

export const SETUP_CLAMS = gql`
  query setupClams {
    setupClams {
      ... on ClamsGameSetup {
        he
      }
    }
  }
`;

export const SETUP_GOAL = gql`
  query setupGoals {
    setupGoals {
      ... on GoalsGameSetup {
        session {
          betId
          betAmount
          currentStep
          selections {
            step
            luckySpots
            selected
          }
          difficulty
          profits {
            step
            multiplier
            profit
          }
          totalProfit {
            multiplier
            profit
          }
          nextProfit {
            multiplier
            profit
          }
          allowNext
          profitCut
        }
        balance
        maxProfit
      }
    }
  }
`;

export const TRANSACTION_BETS = gql`
  query Bets($page: Int, $limit: Int) {
    transactionsBet(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        id
        time
        game
        amount
        multiplier
        profit
      }
    }
  }
`;

export const TRANSACTION_AFFILIATES = gql`
  query AffiliateClaims($page: Int, $limit: Int) {
    transactionsAffiliate(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        id
        claimedAt
        amount
      }
    }
  }
`;

export const TRANSACTION_WITHDRAWALS = gql`
  query transactionsWithdraw($page: Int, $limit: Int) {
    transactionsWithdraw(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        id
        status
        time
        hash
        address
        amount
        fee
      }
    }
  }
`;

export const TRANSACTION_BONUSES = gql`
  query transactionsBonus($page: Int, $limit: Int) {
    transactionsBonus(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        id
        givenAt
        claimedAt
        expiresAt
        amount
        position
        type
        wager
      }
    }
  }
`;

export const FAUCET_INFO = gql`
  query faucetInfo {
    faucetInfo {
      ... on FaucetInfo {
        amount
        every
        canClaim
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const SETUP_MINES = gql`
  query setupMines {
    setupMines {
      ... on MinesGameSetup {
        session {
          betId
          betAmount
          mines
          open
          totalProfit {
            multiplier
            profit
          }
          nextProfit {
            multiplier
            profit
          }
          allowNext
          profitCut
        }
        balance
        maxProfit
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const BONUSCLAIMS = gql`
  query bonusClaims {
    bonusClaims {
      ... on BonusClaim {
        id
        givenAt
        claimedAt
        expiresAt
        amount
        position
        type
        wager
      }
    }
  }
`;
