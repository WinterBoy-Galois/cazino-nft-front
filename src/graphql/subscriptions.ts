import gql from 'graphql-tag';

export const LEADERBOARDS_SUBSCRIPTION = gql`
  subscription leaderboardChanged {
    leaderboardChanged {
      daily {
        username
        userid
        wager
        bonus
      }
      weekly {
        username
        userid
        wager
        bonus
      }
      monthly {
        username
        userid
        wager
        bonus
      }
    }
  }
`;

export const BET_ADDED = gql`
  subscription onBetAdded {
    betAdded {
      id
      time
      userid
      username
      gameid
      bet
      profit
      multiplier
    }
  }
`;

export const BALANCE = gql`
  subscription balance {
    balance {
      user {
        id
        balance
      }
      amount
      event
    }
  }
`;
