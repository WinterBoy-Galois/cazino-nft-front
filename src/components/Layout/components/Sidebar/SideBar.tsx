import React, { useState } from 'react';
import styles from './SideBar.module.scss';
import { useStateValue } from '../../../../state';
import { CSSTransition } from 'react-transition-group';
import SidebarToggle from '../SidebarToggle';
import { useScrollLock } from '../../../../hooks/useScrollLock.hook';
import { Breakpoint, useBreakpoint } from '../../../../hooks/useBreakpoint.hook';
import { SidebarTab } from '../../../../state/models/sidebar.model';
import TabSelect from './components/TabSelect/TabSelect';
import LeaderboardsTab from './components/LeaderboardsTab';
import { useTranslation } from 'react-i18next';
import { useScrollbarWidth } from '../../../../hooks/useScrollbarWidth.hook';
import LatestBetsTab from './components/LatestBetsTab';
import MyBetsTab from './components/MyBetsTab';
import Bet from '../../../../models/bet.model';
import { useBetBuffer, DispatchSpeed } from '../../../../hooks/useBetBuffer.hook';
import { RECENT_BETS } from '../../../../graphql/queries';
import { BET_ADDED } from '../../../../graphql/subscriptions';
import { ApolloError, useQuery, useSubscription } from '@apollo/client';
import { GameTypes } from '../../../../models/gameTypes.model';

const activateScrollLock = (breakpoint: Breakpoint): boolean => {
  switch (breakpoint) {
    case 'xs':
    case 'sm':
      return true;

    default:
      return false;
  }
};

const mapGameType = (gameid: string): GameTypes => {
  const games: { [key: string]: GameTypes } = {
    DICE: GameTypes.DICE,
    GOALS: GameTypes.GOALS,
    MINES: GameTypes.MINES,
    CLAMS: GameTypes.CLAMS,
  };

  return games[gameid];
};

const renderTab = (
  tab: SidebarTab,
  latestBets: Bet[],
  myBets: Bet[],
  isLoading: boolean,
  error: ApolloError | undefined,
  isSignedIn?: boolean
) => {
  switch (tab) {
    case 'LATEST_BETS':
      return <LatestBetsTab bets={latestBets} isLoading={isLoading} error={error} />;
    case 'MY_BETS':
      return (
        <MyBetsTab bets={myBets} isLoading={isLoading} error={error} isSignedIn={isSignedIn} />
      );
    case 'LEADERBOARDS':
      return <LeaderboardsTab />;
  }
};

const SideBar: React.FC = () => {
  const [
    {
      sidebar: { isOpen, selectedTab },
      modal: { type: modalType },
      auth: { state, user },
    },
  ] = useStateValue();
  const breakpoint = useBreakpoint();
  useScrollLock(isOpen && activateScrollLock(breakpoint), false);
  const { t } = useTranslation(['sidebar']);
  const scrollbarWidth = useScrollbarWidth();

  const [latestBets, setLatestBets] = useState<Bet[]>([]);

  const [myBets, setMyBets] = useState<Bet[]>([]);

  const handleBetAdded = (betAdded: Bet) => {
    const newBets = [betAdded, ...latestBets.slice(0, 9)];
    setLatestBets(newBets);
  };

  const handleAddedForCurrentUser = (betAdded: Bet) => {
    const newBets = [betAdded, ...myBets.slice(0, 9)];
    setMyBets(newBets);
  };

  const { addBets } = useBetBuffer({
    bufferSize: 100,
    dispatchSpeed: DispatchSpeed.AUTO,
    currentUserId: user?.id ? +user?.id : 0,
    onBetDispatched: handleBetAdded,
    onBetAddedForCurrentUser: handleAddedForCurrentUser,
  });

  useSubscription(BET_ADDED, {
    onSubscriptionData: data => {
      const betAdded: Bet = data?.subscriptionData?.data?.betAdded;
      if (betAdded) {
        addBets([
          {
            ...betAdded,
            gameid: betAdded.gameid ? mapGameType(betAdded.gameid.toString()) : betAdded.gameid,
          },
        ]);
      }
    },
  });

  const { loading, error } = useQuery(RECENT_BETS, {
    onCompleted: data => {
      if (data?.recentBets?.allBets) {
        const initialLatestBets = data.recentBets.allBets.map((bet: Bet) => {
          return {
            ...bet,
            gameid: bet.gameid ? mapGameType(bet.gameid.toString()) : bet.gameid,
          };
        });
        setLatestBets(initialLatestBets);
      }

      if (data?.recentBets?.myBets) {
        const initialMyBets = data.recentBets.myBets.map((bet: Bet) => {
          return {
            ...bet,
            gameid: bet.gameid ? mapGameType(bet.gameid.toString()) : bet.gameid,
          };
        });
        setMyBets(initialMyBets);
      }
    },
  });

  return (
    <CSSTransition
      in={isOpen}
      timeout={200}
      classNames={{
        enter: styles['slide--enter'],
        enterActive: styles['slide--enter-active'],
        enterDone: styles['slide--enter-active'],
        exit: styles['slide--exit'],
        exitActive: styles['slide--exit-active'],
        exitDone: styles['slide--exit-active'],
      }}
      mountOnEnter={true}
    >
      <div
        className={`${styles.container} ${styles.slide}`}
        style={{ right: `${modalType !== 'NONE' ? `${scrollbarWidth}px` : ''}` }}
      >
        <div className={styles.header}>
          <SidebarToggle show={true} />
          <h1 className={styles['header__headline']}>{t('headline')}</h1>
        </div>
        <div className={styles['tab-select']}>
          <TabSelect />
        </div>
        <div className={`container ${styles['tab-container']}`}>
          {renderTab(selectedTab, latestBets, myBets, loading, error, state === 'SIGNED_IN')}
        </div>
      </div>
    </CSSTransition>
  );
};

export default SideBar;
