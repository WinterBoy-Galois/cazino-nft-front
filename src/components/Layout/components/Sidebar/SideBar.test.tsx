import React from 'react';
import { render } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import SideBar from './SideBar';
import { BET_ADDED } from '../../../../graphql/subscriptions';
import { RECENT_BETS } from '../../../../graphql/queries';

const mocks = [
  {
    request: {
      query: RECENT_BETS,
    },
    result: {
      data: {
        recentBets: {
          allBets: [
            {
              id: '816380',
              time: 1598013566307,
              userid: 478,
              username: 'williamsjohn',
              gameid: 'GOALS',
              bet: 0.00006406,
              profit: 0.00013324,
              multiplier: 3.08,
            },
          ],
          myBets: null,
        },
      },
    },
  },
  {
    request: {
      query: BET_ADDED,
    },
    result: {
      data: {
        betAdded: {
          id: '155689',
          time: 1589971668258,
          userid: 59,
          username: 'pamela56',
          gameid: 'CLAMS',
          bet: 0.00001028,
          profit: -0.00001028,
          multiplier: 0,
        },
      },
    },
  },
];

describe('SideBar', () => {
  it('should match snapshot', async () => {
    // Arrange

    // Act
    const container = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SideBar />
      </MockedProvider>
    );

    // Assert
    expect(container).toMatchSnapshot();
  });
});
