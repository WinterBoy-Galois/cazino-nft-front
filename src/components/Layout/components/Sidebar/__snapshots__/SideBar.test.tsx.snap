// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container slide"
      >
        <div
          class="header"
        >
          <button
            class="button fade"
            style="transform: rotate(0deg);"
          >
            <svg
              class="icon"
              viewBox="0 0 12 26"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <style>
                  .cls-1{fill:none;stroke:#f4feff;stroke-linecap:round;stroke-width:3px}
                </style>
              </defs>
              <g
                data-name="Layer 2"
                id="Layer_2"
              >
                <g
                  data-name="Layer 1"
                  id="Layer_1-2"
                >
                  <g
                    data-name="Group 197"
                    id="Group_197"
                  >
                    <path
                      class="cls-1"
                      d="M1.5 24.5l9-11"
                      data-name="Line 23"
                      id="Line_23"
                    />
                    <path
                      class="cls-1"
                      d="M1.5 1.5l9 12"
                      data-name="Line 24"
                      id="Line_24"
                    />
                  </g>
                </g>
              </g>
            </svg>
          </button>
          <h1
            class="header__headline"
          >
            LIVE UPDATES
          </h1>
        </div>
        <div
          class="tab-select"
        >
          <button
            class="button button--active"
          >
            Latest Bets
          </button>
          <button
            class="button false"
          >
            My Bets
          </button>
          <button
            class="button false"
          >
            Leaderboards
          </button>
        </div>
        <div>
          <div
            class="container"
          >
            <div
              class="bet-table__wrapper"
            >
              <table
                class="bet-table"
              >
                <thead
                  class="bet-table__header"
                >
                  <tr>
                    <th />
                    <th>
                      Time
                    </th>
                    <th>
                      User
                    </th>
                    <th>
                      Profit
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="bet-table__body"
                >
                  <tr>
                    <td>
                      <div />
                    </td>
                    <td
                      class="hide--small hide--medium"
                    >
                      <div />
                    </td>
                    <td
                      class="hide--small hide--medium"
                    >
                      <div />
                    </td>
                    <td>
                      <div />
                    </td>
                  </tr>
                  <p>
                    No Data
                  </p>
                  <tr>
                    <td>
                      <div />
                    </td>
                    <td
                      class="hide--small hide--medium"
                    >
                      <div />
                    </td>
                    <td
                      class="hide--small hide--medium"
                    >
                      <div />
                    </td>
                    <td>
                      <div />
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container slide"
    >
      <div
        class="header"
      >
        <button
          class="button fade"
          style="transform: rotate(0deg);"
        >
          <svg
            class="icon"
            viewBox="0 0 12 26"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <style>
                .cls-1{fill:none;stroke:#f4feff;stroke-linecap:round;stroke-width:3px}
              </style>
            </defs>
            <g
              data-name="Layer 2"
              id="Layer_2"
            >
              <g
                data-name="Layer 1"
                id="Layer_1-2"
              >
                <g
                  data-name="Group 197"
                  id="Group_197"
                >
                  <path
                    class="cls-1"
                    d="M1.5 24.5l9-11"
                    data-name="Line 23"
                    id="Line_23"
                  />
                  <path
                    class="cls-1"
                    d="M1.5 1.5l9 12"
                    data-name="Line 24"
                    id="Line_24"
                  />
                </g>
              </g>
            </g>
          </svg>
        </button>
        <h1
          class="header__headline"
        >
          LIVE UPDATES
        </h1>
      </div>
      <div
        class="tab-select"
      >
        <button
          class="button button--active"
        >
          Latest Bets
        </button>
        <button
          class="button false"
        >
          My Bets
        </button>
        <button
          class="button false"
        >
          Leaderboards
        </button>
      </div>
      <div>
        <div
          class="container"
        >
          <div
            class="bet-table__wrapper"
          >
            <table
              class="bet-table"
            >
              <thead
                class="bet-table__header"
              >
                <tr>
                  <th />
                  <th>
                    Time
                  </th>
                  <th>
                    User
                  </th>
                  <th>
                    Profit
                  </th>
                </tr>
              </thead>
              <tbody
                class="bet-table__body"
              >
                <tr>
                  <td>
                    <div />
                  </td>
                  <td
                    class="hide--small hide--medium"
                  >
                    <div />
                  </td>
                  <td
                    class="hide--small hide--medium"
                  >
                    <div />
                  </td>
                  <td>
                    <div />
                  </td>
                </tr>
                <p>
                  No Data
                </p>
                <tr>
                  <td>
                    <div />
                  </td>
                  <td
                    class="hide--small hide--medium"
                  >
                    <div />
                  </td>
                  <td
                    class="hide--small hide--medium"
                  >
                    <div />
                  </td>
                  <td>
                    <div />
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
